File: 2016/Day11/main.lua
local function create_state(elevator, items, steps)
    return {
        elevator = elevator,
        items = items,
        steps = steps,

        get_hash = function (self)
            local hash = tostring(self.elevator) .. '-'

            -- Create a copy of items to sort
            local sorted_items = {}
            for i, item in ipairs(self.items) do
                sorted_items[i] = {item[1], item[2]}
            end

            -- Sort the items array
            table.sort(sorted_items, function (a, b)
                if a[1] == b[1] then
                    return a[2] < b[2]
                end
                return a[1] < b[1]
            end)

            for _, item in ipairs(sorted_items) do
                hash = hash .. tostring(item[1]) .. ',' .. tostring(item[2]) .. ';'
            end

            return hash
        end,

        is_valid = function (self)
            for i, item in ipairs(self.items) do
                local chip_floor = item[2]

                -- If generator and chip are on the same floor, the chip is safe
                if chip_floor == item[1] then
                    goto continue
                end

                -- Check if any other generator is on the same floor as this chip
                for j, otherItem in ipairs(self.items) do
                    if j ~= i and otherItem[1] == chip_floor then
                        return false
                    end
                end

                ::continue::
            end

            return true
        end,

        is_goal = function (self)
            for _, item in ipairs(self.items) do
                if item[1] ~= 4 or item[2] ~= 4 then
                    return false
                end
            end
            return true
        end
    }
end

local function get_next_states(current)
    local next_states = {}
    local possible_floors = {}

    if current.elevator < 4 then
        table.insert(possible_floors, current.elevator + 1)
    end
    if current.elevator > 1 then
        table.insert(possible_floors, current.elevator - 1)
    end

    local items_on_floor = {} -- {index,  type} where type: 1=gen, 2=chip
    for i, item in ipairs(current.items) do
        if item[1] == current.elevator then
            table.insert(items_on_floor, {i ,1})
        end
        if item[2] == current.elevator then
            table.insert(items_on_floor, {i ,2})
        end
    end

    for _, item1 in ipairs(items_on_floor) do
        -- Try moving one item
        for _, next_floor in ipairs(possible_floors) do
            local new_state = create_state(next_floor, {}, current.steps + 1)

            -- Copy items
            for i, item in ipairs(current.items) do
                new_state.items[i] = {item[1], item[2]}
            end

            -- Upddate the moved item
            if item1[2] == 1 then
                new_state.items[item1[1]][1] = next_floor
            else
                new_state.items[item1[1]][2] = next_floor
            end

            if new_state:is_valid() then
                table.insert(next_states, new_state)
            end
        end

        -- Try moving two items
        for _, item2 in ipairs(items_on_floor) do
            if item1[1] == item2[1] and item1[2] == item2[2] then
                goto continue
            end

            for _, next_floor in ipairs(possible_floors) do
                local new_state = create_state(next_floor, {}, current.steps + 1)

                -- Copy items
                for i, item in ipairs(current.items) do
                    new_state.items[i] = {item[1], item[2]}
                end

                -- Upddate the moved item
                if item1[2] == 1 then
                    new_state.items[item1[1]][1] = next_floor
                else
                    new_state.items[item1[1]][2] = next_floor
                end

                if item2[2] == 1 then
                    new_state.items[item2[1]][1] = next_floor
                else
                    new_state.items[item2[1]][2] = next_floor
                end

                if new_state:is_valid() then
                    table.insert(next_states, new_state)
                end
            end

            ::continue::
        end
    end

    return next_states
end


local function parse_input(input, part2)
    local element_index = {}
    local items = {} -- {generator_floor, chip_floor}

    for floor = 1, #input do
        local line = input[floor]

        -- Find generators
        for element in string.gmatch(line, '(%w+) generator') do
            if not element_index[element] then
                element_index[element] = #items + 1
                items[element_index[element]] = {floor, 0}
            else
                items[element_index[element]][1] = floor
            end
        end

        -- Find microchips
        for element in string.gmatch(line, '(%w+)-compatible microchip') do
            if not element_index[element] then
                element_index[element] = #items + 1
                items[element_index[element]] = {0, floor}
            else
                items[element_index[element]][2] = floor
            end
        end
    end

    if part2 then
        table.insert(items, {1, 1}) -- Elerium generator and microchip
        table.insert(items, {1, 1}) -- Dilithium generator and microchip
    end

    return create_state(1, items, 0)
end

local function find_minimum_steps(initial_state)
    local queue = {initial_state}
    local visited = {}
    visited[initial_state:get_hash()] = true

    local front = 1
    while front <= #queue do
        local current = queue[front]
        front = front + 1

        if current:is_goal() then
            return current.steps
        end

        for _, next_state in ipairs(get_next_states(current)) do
            local hash = next_state:get_hash()
            if not visited[hash] then
                visited[hash] = true
                table.insert(queue, next_state)
            end
        end
    end

    return -1
end

local input = {}
for line in io.lines('input.txt') do
    table.insert(input, line)
end

local initial_state = parse_input(input)
print('Minimum steps: ' .. find_minimum_steps(initial_state))

local initial_state2 = parse_input(input, true)
print('Minimum steps with additional items: ' .. find_minimum_steps(initial_state2))

File: 2016/Day11/main.dart
// Represent a pair, similar to std::pair in C++
// Includes comparison logic for sorting
import 'dart:collection';
import 'dart:io';

class Pair<T1 extends Comparable, T2 extends Comparable>
    implements Comparable<Pair<T1, T2>> {
  T1 first;
  T2 second;

  Pair(this.first, this.second);

  @override
  int compareTo(Pair<T1, T2> other) {
    int cmp = first.compareTo(other.first);
    if (cmp != 0) return cmp;
    return second.compareTo(other.second);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Pair &&
          runtimeType == other.runtimeType &&
          first == other.first &&
          second == other.second;

  @override
  int get hashCode => first.hashCode ^ second.hashCode;

  @override
  String toString() => '($first, $second)';
}

class State {
  final int elevator; // Current floor of the elevator
  // List of pairs: (generator_floor, chip_floor) for each element type
  // The index identifies the element type
  final List<Pair<int, int>> items;
  final int steps; // Number of steps taken to reach this state

  State(this.elevator, List<Pair<int, int>> items, this.steps)
    // Ensure immutability of items if needed,
    // although direct use is fine if we always create new lists for new states
    : items = List.unmodifiable(items);

  // Generates a unique hash for the state
  // Sorts items first so that the orders of elements doesn't affect the hash
  // This is important for the visited set
  String getHash() {
    var sortedItems = List<Pair<int, int>>.from(items);
    sortedItems.sort();

    var hash = StringBuffer();
    hash.write('$elevator-');
    for (final item in sortedItems) {
      hash.write('${item.first},${item.second};');
    }
    return hash.toString();
  }

  // Checks if the current state is valid according to the rules
  // A chip is fried if it's on the same floor as a generator of a *different* type,
  // and its own generator is not on the same floor
  bool isValid() {
    for (int i = 0; i < items.length; i++) {
      int chipFloor = items[i].second;
      int generatorFloor = items[i].first;

      // If chip is safe with its own generator, skip
      if (chipFloor == generatorFloor) continue;

      // Check if any *other* generator is on the same floor as the chip
      for (int j = 0; j < items.length; j++) {
        if (items[j].first == chipFloor) {
          // Found a generator on the same floor as the chip
          // Since we already know chipFloor != generatorFloor (from above),
          // this means the chip is fried
          return false;
        }
      }
    }

    // No fried chips found
    return true;
  }

  // Check if the goal state is reached (all items on floor 4)
  bool isGoal() {
    for (final item in items) {
      if (item.first != 4 || item.second != 4) return false;
    }
    return true;
  }

  // Use the hash representation for use in the visited set
  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is State &&
          runtimeType == other.runtimeType &&
          getHash() == other.getHash();

  @override
  int get hashCode => getHash().hashCode;
}

// Generate all valid next states reachable from the current state
List<State> getNextStates(State current) {
  List<State> nextStates = [];
  List<int> possibleFloors = [];

  // Determine possible floors (up or down)
  if (current.elevator < 4) possibleFloors.add(current.elevator + 1);
  if (current.elevator > 1) possibleFloors.add(current.elevator - 1);

  // Identify items and their types currently on the same floor as the elevator
  // Stora as (item_index, type), where type 0=generator, 1=chip
  List<Pair<int, int>> itemsOnFloor = [];
  for (int i = 0; i < current.items.length; i++) {
    if (current.items[i].first == current.elevator) {
      itemsOnFloor.add(Pair(i, 0)); // Generator
    }
    if (current.items[i].second == current.elevator) {
      itemsOnFloor.add(Pair(i, 1)); // Chip
    }
  }

  // Iterate through all possible floors
  for (int nextFloor in possibleFloors) {
    // Try moving one item
    for (int i = 0; i < itemsOnFloor.length; i++) {
      final itemToMove1 = itemsOnFloor[i];
      // Create a mutable copy of the current items list
      var nextItems = List<Pair<int, int>>.from(current.items);

      // Update the position of the moved item
      int itemIndex1 = itemToMove1.first;
      int itemType1 = itemToMove1.second;
      if (itemType1 == 0) {
        // Move generator
        nextItems[itemIndex1] = Pair(nextFloor, nextItems[itemIndex1].second);
      } else {
        // Move chip
        nextItems[itemIndex1] = Pair(nextItems[itemIndex1].first, nextFloor);
      }

      // Create the potential next state
      var newState = State(nextFloor, nextItems, current.steps + 1);

      // Add to list if valid
      if (newState.isValid()) nextStates.add(newState);
    }

    // Try moving two items
    for (int i = 0; i < itemsOnFloor.length; i++) {
      for (int j = i + 1; j < itemsOnFloor.length; j++) {
        final itemToMove1 = itemsOnFloor[i];
        final itemToMove2 = itemsOnFloor[j];

        // Create a mutable copy of the current items list
        var nextItems = List<Pair<int, int>>.from(current.items);

        int itemIndex1 = itemToMove1.first;
        int itemType1 = itemToMove1.second;
        if (itemType1 == 0) {
          // Move generator
          nextItems[itemIndex1] = Pair(nextFloor, nextItems[itemIndex1].second);
        } else {
          // Move chip
          nextItems[itemIndex1] = Pair(nextItems[itemIndex1].first, nextFloor);
        }

        int itemIndex2 = itemToMove2.first;
        int itemType2 = itemToMove2.second;
        if (itemType2 == 0) {
          // Move generator
          nextItems[itemIndex2] = Pair(nextFloor, nextItems[itemIndex2].second);
        } else {
          // Move chip
          nextItems[itemIndex2] = Pair(nextItems[itemIndex2].first, nextFloor);
        }

        // Create the potential next state
        var newState = State(nextFloor, nextItems, current.steps + 1);

        // Add to list if valid
        if (newState.isValid()) nextStates.add(newState);
      }
    }
  }

  return nextStates;
}

State parseInput(List<String> input, {bool part2 = false}) {
  Map<String, int> elementIndex = {};
  List<int?> generatorFloors = [];
  List<int?> chipFloors = [];
  List<String> elementNames = []; // To keep track of order

  final genRegex = RegExp(r'(\w+) generator');
  final chipRegex = RegExp(r'(\w+)-compatible microchip');

  for (int floor = 0; floor < input.length; floor++) {
    String line = input[floor];
    int currentFloor = floor + 1; // Floors are 1-based

    // Find generators
    for (final match in genRegex.allMatches(line)) {
      String element = match.group(1)!;
      if (!elementIndex.containsKey(element)) {
        elementIndex[element] = elementIndex.length;
        elementNames.add(element);
        // Add placeholders
        generatorFloors.add(null);
        chipFloors.add(null);
      }
      generatorFloors[elementIndex[element]!] = currentFloor;
    }

    // Find chips
    for (final match in chipRegex.allMatches(line)) {
      String element = match.group(1)!;
      if (!elementIndex.containsKey(element)) {
        elementIndex[element] = elementIndex.length;
        elementNames.add(element);
        // Add placeholders
        generatorFloors.add(null);
        chipFloors.add(null);
      }
      chipFloors[elementIndex[element]!] = currentFloor;
    }
  }

  // Final items list from the parsed floors
  List<Pair<int, int>> items = [];
  for (int i = 0; i < elementNames.length; i++) {
    // Assume valid input means both generator and chip are found eventually
    items.add(Pair(generatorFloors[i]!, chipFloors[i]!));
  }

  if (part2) {
    items.add(Pair(1, 1)); // Add elerium generator and chip on floor 1
    items.add(Pair(1, 1)); // Add dilithium generator and chip on floor 1
  }

  return State(1, items, 0);
}

// Find the minimum number of steps using BFS
int findMinimumSteps(State initialState) {
  Queue<State> queue = Queue();
  Set<String> visited = {};

  queue.add(initialState);
  visited.add(initialState.getHash());

  while (queue.isNotEmpty) {
    State current = queue.removeFirst();

    if (current.isGoal()) return current.steps;

    // Explore next possible states
    for (final nextState in getNextStates(current)) {
      String hash = nextState.getHash();
      if (!visited.contains(hash)) {
        visited.add(hash);
        queue.add(nextState);
      }
    }
  }

  return -1; // If no solution found
}

void main() async {
  final lines = await File('input.txt').readAsLines();

  // Part 1
  State initialState = parseInput(lines);
  int result = findMinimumSteps(initialState);
  print('Minimum steps: $result');

  // Part 2
  State initialState2 = parseInput(lines, part2: true);
  int result2 = findMinimumSteps(initialState2);
  print('Minimum steps with additional items: $result2');
}
